### PR TITLE
Refactor authentication logic to conform to strategic auth design.

### PR DESCRIPTION
**Pull request by mkistler**
This PR refactors the old "authenticationMethod" class into a new class named "Authenticator" following the strategic cross language authentication design.

### CODE REVIEW

**Comment by jeff-arn** [comment]
@mkistler it looks like the linux tests are failing, and the OSX image is missing the `TestCredentials.swift` file

**Review by jeff-arn — APPROVED** [review]
This looks really good! Approved.

One thing I want to ask about is whether or not RestKit is using automated SemVer releases? If we end up folding this into the swift SDK it won't matter but we want to make sure this gets released on a major release number if it is not automated.

**Review by mediumTaj — APPROVED** [review]
👍 looks good!

File: Sources/RestKit/Authentication/Authentication.swift [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+    public func authenticate(request: URLRequest, completionHandler: @escaping (URLRequest?, RestError?) -> Void) {
+        var request = request
+        request.addValue("Basic \(encodedCredentials)", forHTTPHeaderField: "Authorization")
+        completionHandler(request, nil)
+    }
+}
+
+/** Authenticate with a static bearer token. */
+public class BearerAuthenticator: Authenticator {
```
Comment (padamstx) [inline Sources/RestKit/Authentication/Authentication.swift]: Design doc calls this `BearerTokenAuthenticator`.

File: Sources/RestKit/Authentication/CloudPakForDataAuthenticator.swift [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    public init(username: String, password: String, url: String) {
+        super.init(tokenSource: CloudPakForDataTokenSource(username: username, password: password, url: url))
+    }
+}
+
+class CloudPakForDataTokenSource: TokenSource {
+
+    let url: String
+
+    var requestHeaders: [String: String]?
```
Comment (padamstx) [inline Sources/RestKit/Authentication/CloudPakForDataAuthenticator.swift]: In design doc, these are just called `headers`.

File: Sources/RestKit/Authentication/IAMAuthenticator.swift [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+
+import Foundation
+
+/** Authenticate with an IAM API key. The API key is used to automatically retrieve and refresh access tokens. */
+public class IAMAuthenticator: TokenSourceAuthenticator {
```
Comment (padamstx) [inline Sources/RestKit/Authentication/IAMAuthenticator.swift:20]: In design doc, this is called `IamAuthenticator` :)

File: Sources/RestKit/Authentication/Authentication.swift [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+     Authenticate a `URLRequest`.
+
+     - parameter request: The request that should be authenticated.
+     - parameter completionHandler: The completion handler to execute with the authenticated `URLRequest`.
+     */
+    func authenticate(request: URLRequest, completionHandler: @escaping (URLRequest?, RestError?) -> Void)
+}
+
+/** No authentication. */
+public class NoAuthAuthenticator: Authenticator {
```
Comment (padamstx) [inline Sources/RestKit/Authentication/Authentication.swift:47]: In the design doc, this is called `NoauthAuthenticator` :)

File: Sources/RestKit/Authentication/IAMAuthenticator.swift [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+    public init(apiKey: String, url: String? = nil) {
+        super.init(tokenSource: IAMTokenSource(apiKey: apiKey, url: url))
+    }
+
+    public func setClientCredentials(clientID: String, clientSecret: String) {
```
Comment (padamstx) [inline Sources/RestKit/Authentication/IAMAuthenticator.swift:26]: In the design doc, the client id is called `clientId` :)

**Review by padamstx — APPROVED** [review]
I had only minor comments about names of things not quite matching what's in the design document.
If it's important for RestKit to be consistent in that regard with the other cores, then those changes should be made, otherwise ignore...

**Pull request merged**